// @jackclaw/watchdog/dist/alerts.js
"use strict";
// Watchdog 告警推送到 ClawChat
// 当检测到以下情况时，通过 Hub /api/chat/send 发消息给 CEO：
// - memory hash 异常变化
// - trust score 骤降（>10分）
// - 进程内存超过阈值（>500MB）
// - 任务失败率过高（>30%）
Object.defineProperty(exports, "__esModule", { value: true });
exports.WatchdogAlerter = void 0;
exports.configureAlerter = configureAlerter;
exports.getAlerter = getAlerter;
class WatchdogAlerter {
    constructor(nodeId, hubUrl, ceoNodeId = "ceo") {
        this.nodeId = nodeId;
        this.hubUrl = hubUrl;
        this.ceoNodeId = ceoNodeId;
    }
    async alert(type, detail) {
        const emoji = { "memory-tamper": "🚨", "trust-drop": "⚠️", "memory-oom": "💾", "high-failure": "❌" }[type];
        await fetch(`${this.hubUrl}/api/chat/send`, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                id: `watchdog-${Date.now()}`,
                from: `watchdog-${this.nodeId}`,
                to: this.ceoNodeId,
                content: `${emoji} [Watchdog Alert] ${type}: ${detail}`,
                type: "human",
                createdAt: Date.now(),
                metadata: { alertType: type, severity: type === "memory-tamper" ? "critical" : "warning" }
            })
        }).catch(() => { }); // 告警失败不崩溃
    }
}
exports.WatchdogAlerter = WatchdogAlerter;
// ─── Module-level singleton for use inside monitor.ts ─────────────────────────
let _alerter = null;
/**
 * 初始化模块级 alerter（由 Hub 或 Node 启动时调用一次）
 */
function configureAlerter(nodeId, hubUrl, ceoNodeId) {
    _alerter = new WatchdogAlerter(nodeId, hubUrl, ceoNodeId);
}
/**
 * 获取当前 alerter（未配置时返回 null，调用方做 null check）
 */
function getAlerter() {
    return _alerter;
}
