// @jackclaw/watchdog/dist/audit-log.js
"use strict";
/**
 * AuditLog — Immutable audit trail for JackClaw Hub
 *
 * Records all security-relevant events:
 *   - Auth (login, register, token refresh)
 *   - Messages (send, receive, delete)
 *   - Admin (config change, plugin load/unload)
 *   - Review (human approval/rejection)
 *
 * Append-only, no delete. Supports export for compliance.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.auditLog = exports.AuditLog = void 0;
class AuditLog {
    constructor(opts) {
        this.entries = [];
        this.counter = 0;
        this.MAX_ENTRIES = opts?.maxEntries ?? 10000;
    }
    /**
     * Append an audit entry. Returns the entry with assigned id.
     */
    append(action, actor, opts) {
        const entry = {
            id: ++this.counter,
            action,
            actor,
            target: opts?.target,
            detail: opts?.detail,
            meta: opts?.meta,
            ts: Date.now(),
        };
        this.entries.push(entry);
        if (this.entries.length > this.MAX_ENTRIES) {
            this.entries = this.entries.slice(-Math.floor(this.MAX_ENTRIES / 2));
        }
        return entry;
    }
    /**
     * Query entries by action prefix.
     */
    query(opts) {
        let result = this.entries;
        if (opts?.action) {
            const prefix = opts.action;
            result = result.filter(e => e.action.startsWith(prefix));
        }
        if (opts?.actor) {
            result = result.filter(e => e.actor === opts.actor);
        }
        if (opts?.from) {
            result = result.filter(e => e.ts >= opts.from);
        }
        if (opts?.to) {
            result = result.filter(e => e.ts <= opts.to);
        }
        const limit = opts?.limit ?? 100;
        return result.slice(-limit);
    }
    /**
     * Export all entries as JSON lines (for compliance / backup).
     */
    exportJSONL() {
        return this.entries.map(e => JSON.stringify(e)).join('\n');
    }
    /**
     * Get entry count.
     */
    get size() {
        return this.entries.length;
    }
    /**
     * Get stats summary.
     */
    stats() {
        const byAction = {};
        for (const e of this.entries) {
            const prefix = e.action.split('.')[0];
            byAction[prefix] = (byAction[prefix] ?? 0) + 1;
        }
        return { total: this.entries.length, byAction };
    }
}
exports.AuditLog = AuditLog;
/** Singleton AuditLog */
exports.auditLog = new AuditLog();
