// @jackclaw/watchdog/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.auditLog = exports.AuditLog = exports.anomalyDetector = exports.AnomalyDetector = exports.getAlerter = exports.configureAlerter = exports.WatchdogAlerter = exports.getLatestSnapshot = exports.getAlerts = exports.humanAck = exports.raiseAlert = exports.compareSnapshots = exports.takeSnapshot = exports.getPolicies = exports.removePolicy = exports.addPolicy = exports.canModify = exports.WatchdogMonitor = void 0;
__exportStar(require("./types"), exports);
__exportStar(require("./isolation"), exports);
var monitor_1 = require("./monitor");
Object.defineProperty(exports, "WatchdogMonitor", { enumerable: true, get: function () { return monitor_1.WatchdogMonitor; } });
Object.defineProperty(exports, "canModify", { enumerable: true, get: function () { return monitor_1.canModify; } });
Object.defineProperty(exports, "addPolicy", { enumerable: true, get: function () { return monitor_1.addPolicy; } });
Object.defineProperty(exports, "removePolicy", { enumerable: true, get: function () { return monitor_1.removePolicy; } });
Object.defineProperty(exports, "getPolicies", { enumerable: true, get: function () { return monitor_1.getPolicies; } });
Object.defineProperty(exports, "takeSnapshot", { enumerable: true, get: function () { return monitor_1.takeSnapshot; } });
Object.defineProperty(exports, "compareSnapshots", { enumerable: true, get: function () { return monitor_1.compareSnapshots; } });
Object.defineProperty(exports, "raiseAlert", { enumerable: true, get: function () { return monitor_1.raiseAlert; } });
Object.defineProperty(exports, "humanAck", { enumerable: true, get: function () { return monitor_1.humanAck; } });
Object.defineProperty(exports, "getAlerts", { enumerable: true, get: function () { return monitor_1.getAlerts; } });
Object.defineProperty(exports, "getLatestSnapshot", { enumerable: true, get: function () { return monitor_1.getLatestSnapshot; } });
var alerts_1 = require("./alerts");
Object.defineProperty(exports, "WatchdogAlerter", { enumerable: true, get: function () { return alerts_1.WatchdogAlerter; } });
Object.defineProperty(exports, "configureAlerter", { enumerable: true, get: function () { return alerts_1.configureAlerter; } });
Object.defineProperty(exports, "getAlerter", { enumerable: true, get: function () { return alerts_1.getAlerter; } });
var anomaly_detector_1 = require("./anomaly-detector");
Object.defineProperty(exports, "AnomalyDetector", { enumerable: true, get: function () { return anomaly_detector_1.AnomalyDetector; } });
Object.defineProperty(exports, "anomalyDetector", { enumerable: true, get: function () { return anomaly_detector_1.anomalyDetector; } });
var audit_log_1 = require("./audit-log");
Object.defineProperty(exports, "AuditLog", { enumerable: true, get: function () { return audit_log_1.AuditLog; } });
Object.defineProperty(exports, "auditLog", { enumerable: true, get: function () { return audit_log_1.auditLog; } });
