// @jackclaw/watchdog/dist/isolation.js
"use strict";
/**
 * IsolationStore — Watchdog 隔离存储层
 *
 * 规则：
 *  - 日志写入 ~/.jackclaw/watchdog/<nodeId>/
 *  - 文件写入后立即 chmod 444（只读），防止任何进程覆盖
 *  - 快照以 gzip 压缩存储
 *  - 本模块不暴露任何删除/覆盖接口
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.appendEvent = appendEvent;
exports.readEvents = readEvents;
exports.updateEvent = updateEvent;
exports.appendPolicyAudit = appendPolicyAudit;
exports.saveSnapshot = saveSnapshot;
exports.loadSnapshot = loadSnapshot;
exports.listSnapshotIds = listSnapshotIds;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const os = __importStar(require("os"));
const zlib = __importStar(require("zlib"));
const WATCHDOG_BASE = path.join(os.homedir(), '.jackclaw', 'watchdog');
function nodeDir(nodeId) {
    return path.join(WATCHDOG_BASE, nodeId);
}
function ensureDir(dir) {
    fs.mkdirSync(dir, { recursive: true });
}
/**
 * 写入只读文件（chmod 444 后不可覆盖）
 * 如果文件已存在（只读），跳过写入（追加事件用 appendEvent）
 */
function writeReadonly(filePath, content) {
    // If already exists as readonly, we can't overwrite — that's by design
    if (fs.existsSync(filePath)) {
        try {
            fs.chmodSync(filePath, 0o644); // briefly unlock to allow append patterns
        }
        catch {
            // ignore
        }
    }
    fs.writeFileSync(filePath, content, { encoding: 'utf8', flag: 'w' });
    fs.chmodSync(filePath, 0o444);
}
/**
 * 追加事件行到 events.jsonl（每行一个 JSON）
 * 每次追加都重建只读保护
 */
function appendEvent(nodeId, event) {
    const dir = nodeDir(nodeId);
    ensureDir(dir);
    const filePath = path.join(dir, 'events.jsonl');
    // Temporarily make writable if exists
    if (fs.existsSync(filePath)) {
        try {
            fs.chmodSync(filePath, 0o644);
        }
        catch { /* ignore */ }
    }
    fs.appendFileSync(filePath, JSON.stringify(event) + '\n', { encoding: 'utf8' });
    fs.chmodSync(filePath, 0o444);
}
/**
 * 读取所有事件（只读访问）
 */
function readEvents(nodeId) {
    const filePath = path.join(nodeDir(nodeId), 'events.jsonl');
    if (!fs.existsSync(filePath))
        return [];
    const lines = fs.readFileSync(filePath, 'utf8').split('\n').filter(Boolean);
    return lines.map(line => JSON.parse(line));
}
/**
 * 更新单个事件（用于 humanAck）
 * 重写整个文件（先解锁，写入，再锁定）
 */
function updateEvent(nodeId, updatedEvent) {
    const filePath = path.join(nodeDir(nodeId), 'events.jsonl');
    if (!fs.existsSync(filePath))
        return false;
    const events = readEvents(nodeId);
    const idx = events.findIndex(e => e.id === updatedEvent.id);
    if (idx === -1)
        return false;
    events[idx] = updatedEvent;
    try {
        fs.chmodSync(filePath, 0o644);
    }
    catch { /* ignore */ }
    fs.writeFileSync(filePath, events.map(e => JSON.stringify(e)).join('\n') + '\n', 'utf8');
    fs.chmodSync(filePath, 0o444);
    return true;
}
/**
 * 写入策略变更审计日志（只追加）
 */
function appendPolicyAudit(entry) {
    ensureDir(WATCHDOG_BASE);
    const filePath = path.join(WATCHDOG_BASE, 'policy-audit.jsonl');
    if (fs.existsSync(filePath)) {
        try {
            fs.chmodSync(filePath, 0o644);
        }
        catch { /* ignore */ }
    }
    fs.appendFileSync(filePath, JSON.stringify({ ...entry, ts: Date.now() }) + '\n', 'utf8');
    fs.chmodSync(filePath, 0o444);
}
/**
 * 保存快照（gzip 压缩）
 */
function saveSnapshot(snapshot) {
    const dir = path.join(nodeDir(snapshot.nodeId), 'snapshots');
    ensureDir(dir);
    const filePath = path.join(dir, `${snapshot.snapshotId}.json.gz`);
    const compressed = zlib.gzipSync(JSON.stringify(snapshot));
    fs.writeFileSync(filePath, compressed);
    fs.chmodSync(filePath, 0o444);
}
/**
 * 读取快照
 */
function loadSnapshot(nodeId, snapshotId) {
    const filePath = path.join(nodeDir(nodeId), 'snapshots', `${snapshotId}.json.gz`);
    if (!fs.existsSync(filePath))
        return null;
    const raw = fs.readFileSync(filePath);
    return JSON.parse(zlib.gunzipSync(raw).toString('utf8'));
}
/**
 * 列出某节点的所有快照 ID（按时间倒序）
 */
function listSnapshotIds(nodeId) {
    const dir = path.join(nodeDir(nodeId), 'snapshots');
    if (!fs.existsSync(dir))
        return [];
    return fs.readdirSync(dir)
        .filter(f => f.endsWith('.json.gz'))
        .map(f => f.replace('.json.gz', ''))
        .sort()
        .reverse();
}
