// @jackclaw/watchdog/dist/monitor.js
"use strict";
/**
 * WatchdogMonitor — 监督层核心
 *
 * 设计原则（来自 Messy Jobs 第6章）：
 *  "必须保留推翻 AI 决策的权力，而这个权力不能交给 AI 自己"
 *
 * 关键约束：
 *  - canModify() 始终返回 false，任何 Agent 调用均被拒绝
 *  - 告警只能由真人通过 humanAck 关闭
 *  - 策略变更全部写入隔离审计日志
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WatchdogMonitor = void 0;
exports.canModify = canModify;
exports.addPolicy = addPolicy;
exports.removePolicy = removePolicy;
exports.getPolicies = getPolicies;
exports.takeSnapshot = takeSnapshot;
exports.compareSnapshots = compareSnapshots;
exports.raiseAlert = raiseAlert;
exports.humanAck = humanAck;
exports.getAlerts = getAlerts;
exports.getLatestSnapshot = getLatestSnapshot;
const crypto = __importStar(require("crypto"));
const isolation_1 = require("./isolation");
const alerts_1 = require("./alerts");
// ─── In-memory policy store (policies themselves aren't sensitive logs) ──────
const policies = new Map();
function policyKey(watcherHandle, targetHandle) {
    return `${watcherHandle}::${targetHandle}`;
}
// ─── Public API ───────────────────────────────────────────────────────────────
/**
 * 始终返回 false — Agent 无法通过此检查修改 Watchdog
 */
function canModify() {
    return false;
}
/**
 * 添加监督策略
 * 变更写入隔离审计日志
 */
function addPolicy(policy) {
    const key = policyKey(policy.watcherHandle, policy.targetHandle);
    policies.set(key, policy);
    (0, isolation_1.appendPolicyAudit)({
        action: 'add_policy',
        watcherHandle: policy.watcherHandle,
        targetHandle: policy.targetHandle,
        scope: policy.scope,
        permissions: policy.permissions,
    });
}
/**
 * 移除监督策略
 * 变更写入隔离审计日志
 */
function removePolicy(watcherHandle, targetHandle) {
    const key = policyKey(watcherHandle, targetHandle);
    const existed = policies.has(key);
    if (existed) {
        const policy = policies.get(key);
        policy.revokedAt = Date.now();
        policies.delete(key);
        (0, isolation_1.appendPolicyAudit)({
            action: 'remove_policy',
            watcherHandle,
            targetHandle,
        });
    }
    return existed;
}
/**
 * 获取节点的所有有效策略
 */
function getPolicies(nodeId) {
    return Array.from(policies.values()).filter(p => p.targetHandle === nodeId && !p.revokedAt);
}
/**
 * 拍摄节点状态快照（供后续对比检测异常）
 * tasks / trustScores 由调用方注入（Watchdog 不主动拉取，保持隔离）
 */
function takeSnapshot(nodeId, tasks, memoryContent, trustScores) {
    const snapshot = {
        nodeId,
        snapshotId: crypto.randomUUID(),
        tasks,
        memoryHash: crypto.createHash('sha256').update(memoryContent).digest('hex'),
        trustScores,
        timestamp: Date.now(),
    };
    (0, isolation_1.saveSnapshot)(snapshot);
    return snapshot;
}
/**
 * 对比两个快照，返回检测到的异常事件列表（未写入日志，仅返回）
 * 调用方决定是否 raiseAlert
 */
function compareSnapshots(watcherNodeId, a, b) {
    if (a.nodeId !== b.nodeId) {
        throw new Error('Cannot compare snapshots from different nodes');
    }
    const anomalies = [];
    const nodeId = a.nodeId;
    // Memory hash changed
    if (a.memoryHash !== b.memoryHash) {
        anomalies.push({
            type: 'memory_anomaly',
            severity: 'warning',
            nodeId,
            watcherNodeId,
            description: `Memory hash changed between snapshots ${a.snapshotId} → ${b.snapshotId}`,
            evidence: {
                prevHash: a.memoryHash,
                newHash: b.memoryHash,
                prevSnapshotId: a.snapshotId,
                newSnapshotId: b.snapshotId,
            },
        });
    }
    // Trust score drops
    for (const [peer, prevScore] of Object.entries(a.trustScores)) {
        const newScore = b.trustScores[peer];
        if (newScore !== undefined && newScore < prevScore - 0.2) {
            anomalies.push({
                type: 'trust_drop',
                severity: newScore < 0.3 ? 'critical' : 'warning',
                nodeId,
                watcherNodeId,
                description: `Trust score for peer "${peer}" dropped: ${prevScore.toFixed(2)} → ${newScore.toFixed(2)}`,
                evidence: { peer, prevScore, newScore },
            });
        }
    }
    // Task count sudden drop (potential task_deviation)
    if (b.tasks.length < a.tasks.length * 0.5 && a.tasks.length > 2) {
        anomalies.push({
            type: 'task_deviation',
            severity: 'warning',
            nodeId,
            watcherNodeId,
            description: `Task count dropped significantly: ${a.tasks.length} → ${b.tasks.length}`,
            evidence: { prevCount: a.tasks.length, newCount: b.tasks.length },
        });
    }
    return anomalies;
}
/**
 * 触发告警 — 写入隔离日志，不可被 Agent 删除
 */
function raiseAlert(partial) {
    const event = {
        ...partial,
        id: crypto.randomUUID(),
        timestamp: Date.now(),
        acknowledged: false,
    };
    (0, isolation_1.appendEvent)(event.nodeId, event);
    // Push to ClawChat if alerter is configured
    const alerter = (0, alerts_1.getAlerter)();
    if (alerter) {
        const typeMap = {
            memory_anomaly: "memory-tamper",
            trust_drop: "trust-drop",
        };
        const chatType = typeMap[event.type];
        if (chatType) {
            alerter.alert(chatType, event.description).catch(() => { });
        }
    }
    return event;
}
/**
 * 真人确认告警 — humanId 必须由调用方验证为真人 token
 * Agent 调用此函数时应在路由层被 human-token 中间件拦截
 */
function humanAck(eventId, nodeId, humanId) {
    const events = (0, isolation_1.readEvents)(nodeId);
    const event = events.find(e => e.id === eventId);
    if (!event)
        return false;
    if (event.acknowledged)
        return true; // already acked
    const updated = {
        ...event,
        acknowledged: true,
        acknowledgedBy: humanId,
    };
    return (0, isolation_1.updateEvent)(nodeId, updated);
}
/**
 * 查询告警（支持过滤）
 */
function getAlerts(nodeId, opts = {}) {
    let events = (0, isolation_1.readEvents)(nodeId);
    if (opts.severity !== undefined) {
        events = events.filter(e => e.severity === opts.severity);
    }
    if (opts.acknowledged !== undefined) {
        events = events.filter(e => e.acknowledged === opts.acknowledged);
    }
    if (opts.since !== undefined) {
        events = events.filter(e => e.timestamp >= opts.since);
    }
    // Sort descending by timestamp
    events.sort((a, b) => b.timestamp - a.timestamp);
    if (opts.limit !== undefined) {
        events = events.slice(0, opts.limit);
    }
    return events;
}
/**
 * 获取节点最新快照
 */
function getLatestSnapshot(nodeId) {
    const ids = (0, isolation_1.listSnapshotIds)(nodeId);
    if (ids.length === 0)
        return null;
    return (0, isolation_1.loadSnapshot)(nodeId, ids[0]);
}
// ─── WatchdogMonitor class (OOP facade for HTTP routes) ──────────────────────
class WatchdogMonitor {
    canModify() { return canModify(); }
    addPolicy(policy) { addPolicy(policy); }
    removePolicy(watcherHandle, targetHandle) {
        return removePolicy(watcherHandle, targetHandle);
    }
    takeSnapshot(nodeId, tasks, memoryContent, trustScores) {
        return takeSnapshot(nodeId, tasks, memoryContent, trustScores);
    }
    compareSnapshots(watcherNodeId, a, b) {
        return compareSnapshots(watcherNodeId, a, b);
    }
    raiseAlert(partial) {
        return raiseAlert(partial);
    }
    humanAck(eventId, nodeId, humanId) {
        return humanAck(eventId, nodeId, humanId);
    }
    getAlerts(nodeId, opts) {
        return getAlerts(nodeId, opts);
    }
    getLatestSnapshot(nodeId) {
        return getLatestSnapshot(nodeId);
    }
    getPolicies(nodeId) {
        return getPolicies(nodeId);
    }
}
exports.WatchdogMonitor = WatchdogMonitor;
