// @jackclaw/protocol/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHumanId = exports.isNodeId = exports.isAgentHandle = exports.threadId = exports.hubId = exports.nodeId = exports.brandedAgentHandle = exports.humanId = exports.openMessage = exports.createMessage = void 0;
// @jackclaw/protocol public API
__exportStar(require("./types"), exports);
__exportStar(require("./crypto"), exports);
var message_1 = require("./message");
Object.defineProperty(exports, "createMessage", { enumerable: true, get: function () { return message_1.createMessage; } });
Object.defineProperty(exports, "openMessage", { enumerable: true, get: function () { return message_1.openMessage; } });
__exportStar(require("./identity"), exports);
__exportStar(require("./task-bundle"), exports);
__exportStar(require("./human-in-loop"), exports);
__exportStar(require("./payment-vault"), exports);
__exportStar(require("./social"), exports);
__exportStar(require("./receipt"), exports);
__exportStar(require("./federation"), exports);
__exportStar(require("./concierge"), exports);
// Unified identity model (branded types) — use UI_ prefix to avoid conflict with identity.ts
var unified_identity_1 = require("./unified-identity");
Object.defineProperty(exports, "humanId", { enumerable: true, get: function () { return unified_identity_1.humanId; } });
Object.defineProperty(exports, "brandedAgentHandle", { enumerable: true, get: function () { return unified_identity_1.agentHandle; } });
Object.defineProperty(exports, "nodeId", { enumerable: true, get: function () { return unified_identity_1.nodeId; } });
Object.defineProperty(exports, "hubId", { enumerable: true, get: function () { return unified_identity_1.hubId; } });
Object.defineProperty(exports, "threadId", { enumerable: true, get: function () { return unified_identity_1.threadId; } });
Object.defineProperty(exports, "isAgentHandle", { enumerable: true, get: function () { return unified_identity_1.isAgentHandle; } });
Object.defineProperty(exports, "isNodeId", { enumerable: true, get: function () { return unified_identity_1.isNodeId; } });
Object.defineProperty(exports, "isHumanId", { enumerable: true, get: function () { return unified_identity_1.isHumanId; } });
