// @jackclaw/protocol/dist/message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMessage = createMessage;
exports.openMessage = openMessage;
const crypto_1 = require("./crypto");
/**
 * Canonical string for signing: deterministic serialisation of message fields
 * excluding the signature itself.
 */
function signingData(msg) {
    return JSON.stringify({
        from: msg.from,
        to: msg.to,
        type: msg.type,
        payload: msg.payload,
        timestamp: msg.timestamp,
    });
}
/**
 * Build and sign an JackClawMessage.
 *
 * @param from           Sender node ID
 * @param to             Recipient node ID
 * @param type           Message type
 * @param payloadObject  Plain object to encrypt (will be JSON-serialised)
 * @param recipientPub   Recipient's RSA public key PEM (for encryption)
 * @param senderPriv     Sender's RSA private key PEM (for signing)
 */
function createMessage(from, to, type, payloadObject, recipientPub, senderPriv) {
    const plaintext = JSON.stringify(payloadObject);
    const encPayload = (0, crypto_1.encrypt)(plaintext, recipientPub);
    const payloadStr = JSON.stringify(encPayload);
    const partial = {
        from,
        to,
        type,
        payload: payloadStr,
        timestamp: Date.now(),
    };
    const signature = (0, crypto_1.sign)(signingData(partial), senderPriv);
    return { ...partial, signature };
}
/**
 * Verify and decrypt an JackClawMessage.
 *
 * @param msg           The received message
 * @param senderPub     Sender's RSA public key PEM (for verification)
 * @param recipientPriv Recipient's RSA private key PEM (for decryption)
 * @returns Decrypted payload as parsed object, or throws on failure
 */
function openMessage(msg, senderPub, recipientPriv) {
    // 1. Verify signature
    const dataToVerify = signingData({
        from: msg.from,
        to: msg.to,
        type: msg.type,
        payload: msg.payload,
        timestamp: msg.timestamp,
    });
    if (!(0, crypto_1.verify)(dataToVerify, msg.signature, senderPub)) {
        throw new Error('JackClaw: message signature verification failed');
    }
    // 2. Decrypt payload
    const encPayload = JSON.parse(msg.payload);
    const plaintext = (0, crypto_1.decrypt)(encPayload, recipientPriv);
    return JSON.parse(plaintext);
}
