// @jackclaw/protocol/dist/payment-vault.js
"use strict";
/**
 * JackClaw Payment Vault — 支付隔离区
 *
 * 设计原则：
 * 1. 完全隔离：即使其他区域被攻陷/中毒，Payment Vault 不受影响
 * 2. 司法合规：内置多司法区 AI Agent 支付法规规则引擎
 * 3. 人工审批：超过阈值的支付必须真人确认（类似 Watchdog humanAck）
 * 4. 不可篡改审计：所有支付操作写入只读日志
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMPLIANCE_RULES = void 0;
exports.COMPLIANCE_RULES = {
    CN: {
        jurisdiction: 'CN',
        autoApproveLimit: 137, // ~1000 CNY
        requireHumanAbove: 685, // ~5000 CNY
        requireAuditTrail: true,
        requireRealNameKYC: true,
        allowedPaymentMethods: ['alipay', 'wechat_pay', 'unionpay'],
        prohibitedCategories: ['gambling', 'crypto', 'adult'],
        aiAgentDisclosure: true,
        maxDailyLimit: 1370, // ~10000 CNY
        cooldownSeconds: 60,
    },
    EU: {
        jurisdiction: 'EU',
        autoApproveLimit: 30, // PSD2 SCA 豁免上限
        requireHumanAbove: 30,
        requireAuditTrail: true,
        requireRealNameKYC: true,
        allowedPaymentMethods: ['sepa', 'card', 'paypal'],
        prohibitedCategories: ['gambling', 'adult'],
        aiAgentDisclosure: true, // GDPR 透明度要求
        maxDailyLimit: 1000,
        cooldownSeconds: 30,
    },
    US: {
        jurisdiction: 'US',
        autoApproveLimit: 500,
        requireHumanAbove: 500,
        requireAuditTrail: true,
        requireRealNameKYC: false, // 低金额豁免
        allowedPaymentMethods: ['ach', 'card', 'wire'],
        prohibitedCategories: ['gambling'],
        aiAgentDisclosure: false,
        maxDailyLimit: 5000,
        cooldownSeconds: 0,
    },
    HK: {
        jurisdiction: 'HK',
        autoApproveLimit: 200, // ~1500 HKD
        requireHumanAbove: 1300, // ~10000 HKD
        requireAuditTrail: true,
        requireRealNameKYC: true,
        allowedPaymentMethods: ['fps', 'octopus', 'card'],
        prohibitedCategories: ['gambling', 'adult'],
        aiAgentDisclosure: true,
        maxDailyLimit: 6500,
        cooldownSeconds: 30,
    },
    SG: {
        jurisdiction: 'SG',
        autoApproveLimit: 150, // ~200 SGD
        requireHumanAbove: 750, // ~1000 SGD
        requireAuditTrail: true,
        requireRealNameKYC: true,
        allowedPaymentMethods: ['paynow', 'card', 'grabpay'],
        prohibitedCategories: ['gambling', 'adult', 'crypto'],
        aiAgentDisclosure: true,
        maxDailyLimit: 3750,
        cooldownSeconds: 60,
    },
    GLOBAL: {
        jurisdiction: 'GLOBAL',
        autoApproveLimit: 10, // 最保守：$10以内自动批准
        requireHumanAbove: 10,
        requireAuditTrail: true,
        requireRealNameKYC: true,
        allowedPaymentMethods: [], // 需手动配置
        prohibitedCategories: ['gambling', 'adult', 'crypto', 'weapons'],
        aiAgentDisclosure: true,
        maxDailyLimit: 100,
        cooldownSeconds: 300,
    },
};
