// @jackclaw/protocol/dist/receipt.js
"use strict";
// ClawChat — message delivery/read receipt types
//
// Unified Message State Machine (6 states):
//
//   accepted → sent → acked → stored → consumed → failed
//
//   accepted  : Hub received and persisted the message
//   sent      : Hub pushed message to target WebSocket
//   acked     : Target node sent delivery_ack back to Hub
//   stored    : Target node wrote message to local store
//   consumed  : Target's handler/UI processed the message
//   failed    : Delivery failed after all retries exhausted
//
// Transitions:
//   → accepted : on Hub saveMessage()
//   → sent     : on ws.send() callback success
//   → acked    : on receiving delivery_ack from target node
//   → stored   : on target node confirming local store write
//   → consumed : on target handler/UI marking message as processed
//   → failed   : on delivery timeout with no ack after retries
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATUS_TRANSITIONS = void 0;
exports.isValidTransition = isValidTransition;
/** Valid state transitions */
exports.STATUS_TRANSITIONS = {
    accepted: ['sent', 'failed'],
    sent: ['acked', 'failed'],
    acked: ['stored', 'failed'],
    stored: ['consumed'],
    consumed: [],
    failed: ['accepted'], // retry resets to accepted
    duplicate: [],
};
function isValidTransition(from, to) {
    return exports.STATUS_TRANSITIONS[from]?.includes(to) ?? false;
}
