// @jackclaw/protocol/dist/unified-identity.js
"use strict";
/**
 * JackClaw Unified Identity Model
 *
 * 6 concepts, strictly separated at the type level:
 *
 *   1. HumanId       — a human user (e.g., "hu_abc123")
 *   2. AgentHandle   — an Agent's global address (e.g., "@alice.jackclaw")
 *   3. NodeId        — a runtime node instance (e.g., "node-7f3a")
 *   4. HubId         — a Hub instance (e.g., "hub-prod-01")
 *   5. ThreadId      — a conversation thread (e.g., "thread-uuid")
 *   6. DeliveryTarget — where to deliver a message (handle or nodeId)
 *
 * These are branded string types — cannot be accidentally mixed at compile time.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.humanId = humanId;
exports.agentHandle = agentHandle;
exports.nodeId = nodeId;
exports.hubId = hubId;
exports.threadId = threadId;
exports.isAgentHandle = isAgentHandle;
exports.isNodeId = isNodeId;
exports.isHumanId = isHumanId;
// ─── Constructors (runtime validation + branding) ─────────────────────────────
function humanId(raw) {
    if (!raw || raw.startsWith('@') || raw.startsWith('node-') || raw.startsWith('hub-')) {
        throw new Error(`Invalid HumanId: "${raw}"`);
    }
    return raw;
}
function agentHandle(raw) {
    const h = raw.startsWith('@') ? raw : `@${raw}`;
    return h;
}
function nodeId(raw) {
    return raw;
}
function hubId(raw) {
    return raw;
}
function threadId(raw) {
    return raw;
}
// ─── Type guards ──────────────────────────────────────────────────────────────
function isAgentHandle(s) {
    return s.startsWith('@');
}
function isNodeId(s) {
    return !s.startsWith('@') && !s.startsWith('hu_') && !s.startsWith('hub-') && !s.startsWith('thread-');
}
function isHumanId(s) {
    return s.startsWith('hu_');
}
