// @jackclaw/payment-vault/dist/isolation.js
"use strict";
/**
 * VaultIsolationStore — Payment Vault 隔离存储层
 *
 * 规则（同 Watchdog isolation 模式）：
 *  - 存储路径 ~/.jackclaw/vault/<nodeId>/，独立于 Watchdog/Memory
 *  - payments.jsonl chmod 444，只追加不可修改
 *  - 本模块不暴露删除/覆盖接口
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.VAULT_BASE_DEFAULT = void 0;
exports.appendPaymentLog = appendPaymentLog;
exports.readPaymentLog = readPaymentLog;
exports.computeDailyTotal = computeDailyTotal;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const os = __importStar(require("os"));
const VAULT_BASE_DEFAULT = path.join(os.homedir(), '.jackclaw', 'vault');
exports.VAULT_BASE_DEFAULT = VAULT_BASE_DEFAULT;
function nodeDir(baseDir, nodeId) {
    return path.join(baseDir, nodeId);
}
function ensureDir(dir) {
    fs.mkdirSync(dir, { recursive: true });
}
// ── Append-only payment log ──────────────────────────────────────────────────
function appendPaymentLog(baseDir, nodeId, entry) {
    const dir = nodeDir(baseDir, nodeId);
    ensureDir(dir);
    const filePath = path.join(dir, 'payments.jsonl');
    // Temporarily unlock if exists
    if (fs.existsSync(filePath)) {
        try {
            fs.chmodSync(filePath, 0o644);
        }
        catch { /* ignore */ }
    }
    fs.appendFileSync(filePath, JSON.stringify(entry) + '\n', 'utf8');
    fs.chmodSync(filePath, 0o444);
}
function readPaymentLog(baseDir, nodeId) {
    const filePath = path.join(nodeDir(baseDir, nodeId), 'payments.jsonl');
    if (!fs.existsSync(filePath))
        return [];
    const lines = fs.readFileSync(filePath, 'utf8').split('\n').filter(Boolean);
    return lines.map(line => JSON.parse(line));
}
// ── Daily totals (computed from log) ─────────────────────────────────────────
function computeDailyTotal(baseDir, nodeId) {
    const todayStart = new Date();
    todayStart.setHours(0, 0, 0, 0);
    const todayMs = todayStart.getTime();
    const entries = readPaymentLog(baseDir, nodeId);
    return entries
        .filter(e => e.createdAt >= todayMs &&
        (e.status === 'executed' || e.status === 'approved' || e.status === 'pending_human' || e.status === 'pending_compliance'))
        .reduce((sum, e) => sum + e.amount, 0);
}
