// @jackclaw/payment-vault/dist/vault.js
"use strict";
/**
 * PaymentVault — 隔离支付区核心类
 *
 * 状态机：pending_compliance → approved/pending_human → approved → executed
 *                            → rejected (compliance fail or human reject)
 *
 * 安全约束：
 * - Agent 不能直接执行支付，只能 submit → 合规检查 → 可能等人工审批 → execute
 * - execute() 内部验证状态机，只有 approved 状态才能执行
 * - 人工审批使用 HMAC-SHA256（同 Watchdog/HumanInLoop）
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentVault = void 0;
const crypto_1 = require("crypto");
const protocol_1 = require("@jackclaw/protocol");
const isolation_1 = require("./isolation");
class PaymentVault {
    config;
    baseDir;
    isSandboxMode;
    /** In-memory index of active requests (persisted to JSONL on every mutation) */
    requests = new Map();
    /** Per-node last-payment timestamp for cooldown enforcement */
    lastPaymentAt = new Map();
    constructor(config) {
        this.config = config;
        this.baseDir = config.vaultDir || isolation_1.VAULT_BASE_DEFAULT;
        this.isSandboxMode = process.env.PAYMENT_SANDBOX === 'true';
    }
    warnSandbox() {
        if (this.isSandboxMode) {
            console.warn('[payment] SANDBOX MODE — no real payments');
        }
    }
    // ── Compliance Engine ────────────────────────────────────────────────────────
    checkCompliance(req) {
        const rule = protocol_1.COMPLIANCE_RULES[req.jurisdiction] ?? protocol_1.COMPLIANCE_RULES['GLOBAL'];
        const violations = [];
        // 1. Prohibited category
        if (rule.prohibitedCategories.includes(req.category)) {
            violations.push(`Category "${req.category}" is prohibited in ${req.jurisdiction}`);
        }
        // 2. Payment method whitelist (empty = needs manual config)
        if (rule.allowedPaymentMethods.length > 0 && !rule.allowedPaymentMethods.includes(req.paymentMethod)) {
            violations.push(`Payment method "${req.paymentMethod}" not allowed in ${req.jurisdiction}`);
        }
        if (rule.allowedPaymentMethods.length === 0) {
            violations.push(`No payment methods configured for ${req.jurisdiction} — requires manual setup`);
        }
        // 3. Daily limit
        const dailyTotal = (0, isolation_1.computeDailyTotal)(this.baseDir, req.nodeId);
        if (dailyTotal + req.amount > rule.maxDailyLimit) {
            violations.push(`Daily limit exceeded: current $${dailyTotal} + $${req.amount} > $${rule.maxDailyLimit}`);
        }
        // 4. Cooldown
        const lastTs = this.lastPaymentAt.get(req.nodeId);
        if (lastTs && rule.cooldownSeconds > 0) {
            const elapsed = (Date.now() - lastTs) / 1000;
            if (elapsed < rule.cooldownSeconds) {
                violations.push(`Cooldown: ${Math.ceil(rule.cooldownSeconds - elapsed)}s remaining`);
            }
        }
        // Determine auto-approve or human required
        const requiresHuman = req.amount > rule.autoApproveLimit || req.amount > rule.requireHumanAbove;
        const autoApproved = violations.length === 0 && !requiresHuman;
        return {
            passed: violations.length === 0,
            jurisdiction: req.jurisdiction,
            rule,
            violations,
            requiresHuman,
            autoApproved,
        };
    }
    // ── Submit (Agent entry point) ─────────────────────────────────────────────
    submit(req) {
        this.warnSandbox();
        const requestId = (0, crypto_1.randomUUID)();
        // Run compliance
        const complianceResult = this.checkCompliance({
            amount: req.amount,
            currency: req.currency,
            jurisdiction: req.jurisdiction,
            paymentMethod: req.paymentMethod,
            category: req.category,
            nodeId: req.nodeId,
        });
        let status;
        if (!complianceResult.passed) {
            status = 'rejected';
        }
        else if (complianceResult.autoApproved) {
            status = 'approved';
        }
        else {
            status = 'pending_human';
        }
        const payment = {
            ...req,
            requestId,
            status,
            complianceResult,
            humanApprovalRequired: complianceResult.requiresHuman,
            createdAt: Date.now(),
            auditHash: this.computeAuditHash(requestId, req.amount, req.nodeId),
        };
        this.requests.set(requestId, payment);
        (0, isolation_1.appendPaymentLog)(this.baseDir, req.nodeId, payment);
        return payment;
    }
    // ── Human Approve / Reject ─────────────────────────────────────────────────
    humanApprove(requestId, humanToken) {
        this.warnSandbox();
        const payment = this.getRequest(requestId);
        if (payment.status !== 'pending_human') {
            throw new Error(`Cannot approve: status is "${payment.status}", expected "pending_human"`);
        }
        if (!this.verifyHumanToken(requestId, humanToken)) {
            throw new Error('Invalid human-token. Unauthorized.');
        }
        payment.status = 'approved';
        payment.humanApprovedBy = `human:${humanToken.slice(0, 8)}`;
        payment.humanApprovedAt = Date.now();
        this.requests.set(requestId, payment);
        (0, isolation_1.appendPaymentLog)(this.baseDir, payment.nodeId, payment);
        return payment;
    }
    humanReject(requestId, humanToken, reason) {
        this.warnSandbox();
        const payment = this.getRequest(requestId);
        if (payment.status !== 'pending_human') {
            throw new Error(`Cannot reject: status is "${payment.status}", expected "pending_human"`);
        }
        if (!this.verifyHumanToken(requestId, humanToken)) {
            throw new Error('Invalid human-token. Unauthorized.');
        }
        payment.status = 'rejected';
        payment.failureReason = reason;
        payment.humanApprovedBy = `human:${humanToken.slice(0, 8)}`;
        payment.humanApprovedAt = Date.now();
        this.requests.set(requestId, payment);
        (0, isolation_1.appendPaymentLog)(this.baseDir, payment.nodeId, payment);
        return payment;
    }
    // ── Execute (state-machine enforced) ───────────────────────────────────────
    execute(requestId) {
        this.warnSandbox();
        const payment = this.getRequest(requestId);
        if (payment.status !== 'approved') {
            throw new Error(`Cannot execute: status is "${payment.status}". Only "approved" payments can be executed.`);
        }
        // Verify audit hash integrity
        const expectedHash = this.computeAuditHash(requestId, payment.amount, payment.nodeId);
        if (payment.auditHash !== expectedHash) {
            payment.status = 'failed';
            payment.failureReason = 'Audit hash mismatch — possible tampering';
            this.requests.set(requestId, payment);
            (0, isolation_1.appendPaymentLog)(this.baseDir, payment.nodeId, payment);
            throw new Error('Audit hash verification failed');
        }
        // Execute (log-only in this version — no real payment gateway)
        payment.status = 'executed';
        payment.executedAt = Date.now();
        this.lastPaymentAt.set(payment.nodeId, Date.now());
        this.requests.set(requestId, payment);
        (0, isolation_1.appendPaymentLog)(this.baseDir, payment.nodeId, payment);
        console.log(`[vault] Payment executed: ${requestId} | $${payment.amount} ${payment.currency} → ${payment.recipient} | node=${payment.nodeId}`);
        return payment;
    }
    // ── Read-only queries ──────────────────────────────────────────────────────
    getDailyTotal(nodeId) {
        return (0, isolation_1.computeDailyTotal)(this.baseDir, nodeId);
    }
    getAuditLog(nodeId) {
        return (0, isolation_1.readPaymentLog)(this.baseDir, nodeId);
    }
    getPending() {
        return Array.from(this.requests.values()).filter(r => r.status === 'pending_human');
    }
    getRequest(requestId) {
        const payment = this.requests.get(requestId);
        if (!payment) {
            throw new Error(`Payment request ${requestId} not found`);
        }
        return payment;
    }
    // ── HMAC token (same mechanism as Watchdog/HumanInLoop) ────────────────────
    generateHumanToken(requestId) {
        return (0, crypto_1.createHmac)('sha256', this.config.humanTokenSecret)
            .update(requestId)
            .digest('hex');
    }
    verifyHumanToken(requestId, token) {
        const expected = this.generateHumanToken(requestId);
        try {
            return (0, crypto_1.timingSafeEqual)(Buffer.from(expected, 'hex'), Buffer.from(token, 'hex'));
        }
        catch {
            return false;
        }
    }
    computeAuditHash(requestId, amount, nodeId) {
        return (0, crypto_1.createHmac)('sha256', this.config.humanTokenSecret)
            .update(`${requestId}:${amount}:${nodeId}`)
            .digest('hex');
    }
}
exports.PaymentVault = PaymentVault;
