// @jackclaw/memory/dist/collab.js
"use strict";
// 协作会话 — 零配置 P2P，支持技能转移与教学隔离
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCollabSession = createCollabSession;
const nanoid_1 = require("nanoid");
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const fs_1 = __importDefault(require("fs"));
const SNAPSHOTS_DIR = path_1.default.join(os_1.default.homedir(), '.jackclaw', 'snapshots');
function createCollabSession(opts) {
    const sessionId = (0, nanoid_1.nanoid)();
    const startedAt = Date.now();
    // 独立沙箱，不写 L2
    const sandbox = [];
    const session = {
        id: sessionId,
        intent: opts.intent,
        initiatorId: opts.agentId,
        peerId: opts.peerId,
        topic: opts.topic,
        startedAt,
        share(content, tags = []) {
            const entry = {
                id: (0, nanoid_1.nanoid)(),
                agentId: opts.agentId,
                layer: 'L1',
                category: 'reference',
                scope: `peer:${opts.peerId}`,
                content,
                tags,
                importance: 0.5,
                createdAt: Date.now(),
                updatedAt: Date.now(),
                source: opts.agentId,
            };
            sandbox.push(entry);
            return entry;
        },
        teach(partial) {
            const entry = {
                id: (0, nanoid_1.nanoid)(),
                agentId: opts.agentId,
                layer: 'L1',
                category: (partial.category ?? 'skill'),
                scope: (partial.scope ?? `peer:${opts.peerId}`),
                content: partial.content ?? '',
                tags: partial.tags ?? [],
                importance: partial.importance ?? 0.8,
                createdAt: Date.now(),
                updatedAt: Date.now(),
                source: opts.agentId,
                ...partial,
            };
            sandbox.push(entry);
            return entry;
        },
        async end(mode) {
            // 通知 Hub 结束（如有）
            if (opts.onEnd) {
                await opts.onEnd(sandbox).catch(err => console.error('[collab] onEnd failed:', err));
            }
            switch (mode) {
                case 'discard':
                    // 沙箱完全清空，L2 不变
                    sandbox.length = 0;
                    break;
                case 'archive': {
                    // 沙箱内容写入学习者 L2，source 保留
                    const now = Date.now();
                    for (const entry of sandbox) {
                        opts.store.save({
                            ...entry,
                            layer: 'L2',
                            scope: 'private',
                            agentId: opts.agentId,
                            updatedAt: now,
                        });
                    }
                    sandbox.length = 0;
                    break;
                }
                case 'snapshot': {
                    // 序列化为独立 JSON，不合并主记忆，随时可删/激活
                    fs_1.default.mkdirSync(SNAPSHOTS_DIR, { recursive: true });
                    const snapshotPath = path_1.default.join(SNAPSHOTS_DIR, `teaching-${sessionId}.json`);
                    fs_1.default.writeFileSync(snapshotPath, JSON.stringify({
                        sessionId,
                        intent: opts.intent,
                        peerId: opts.peerId,
                        topic: opts.topic,
                        savedAt: Date.now(),
                        entries: sandbox,
                    }, null, 2));
                    console.log(`[collab] Snapshot saved: ${snapshotPath}`);
                    sandbox.length = 0;
                    break;
                }
                case 'publish':
                    // publish 由 MemoryManager 处理（推送 L3），这里只清沙箱
                    sandbox.length = 0;
                    break;
            }
        },
    };
    return session;
}
