// @jackclaw/memory/dist/index.js
"use strict";
// @jackclaw/memory — 公开 API
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCollabSession = exports.MemDirSync = exports.HubSync = exports.L2Store = exports.L1Cache = exports.MemoryManager = void 0;
// ── 新4分类记忆体系 ──────────────────────────────────────────
var manager_js_1 = require("./manager.js");
Object.defineProperty(exports, "MemoryManager", { enumerable: true, get: function () { return manager_js_1.MemoryManager; } });
// ── 旧三层架构（保留向后兼容） ────────────────────────────────
var l1_cache_js_1 = require("./l1-cache.js");
Object.defineProperty(exports, "L1Cache", { enumerable: true, get: function () { return l1_cache_js_1.L1Cache; } });
var store_js_1 = require("./store.js");
Object.defineProperty(exports, "L2Store", { enumerable: true, get: function () { return store_js_1.L2Store; } });
var sync_js_1 = require("./sync.js");
Object.defineProperty(exports, "HubSync", { enumerable: true, get: function () { return sync_js_1.HubSync; } });
Object.defineProperty(exports, "MemDirSync", { enumerable: true, get: function () { return sync_js_1.MemDirSync; } });
var collab_js_1 = require("./collab.js");
Object.defineProperty(exports, "createCollabSession", { enumerable: true, get: function () { return collab_js_1.createCollabSession; } });
