// @jackclaw/memory/dist/l1-cache.js
"use strict";
// L1 工作记忆 — 内存缓存，会话结束自动清理
Object.defineProperty(exports, "__esModule", { value: true });
exports.L1Cache = void 0;
class L1Cache {
    cache = new Map();
    timer;
    constructor(cleanupIntervalMs = 60_000) {
        this.timer = setInterval(() => this.evictExpired(), cleanupIntervalMs);
    }
    set(entry) {
        this.cache.set(entry.id, entry);
    }
    get(id) {
        const entry = this.cache.get(id);
        if (entry && entry.expiresAt && Date.now() > entry.expiresAt) {
            this.cache.delete(id);
            return undefined;
        }
        return entry;
    }
    query(opts = {}) {
        const now = Date.now();
        const results = [];
        for (const entry of this.cache.values()) {
            if (entry.expiresAt && now > entry.expiresAt)
                continue;
            if (opts.layer && entry.layer !== opts.layer)
                continue;
            if (opts.category && entry.category !== opts.category)
                continue;
            if (opts.scope && entry.scope !== opts.scope)
                continue;
            if (opts.minImportance !== undefined && entry.importance < opts.minImportance)
                continue;
            if (opts.tags?.length) {
                const has = opts.tags.some(t => entry.tags.includes(t));
                if (!has)
                    continue;
            }
            results.push(entry);
        }
        return results
            .sort((a, b) => b.importance - a.importance)
            .slice(0, opts.limit ?? 50);
    }
    delete(id) {
        this.cache.delete(id);
    }
    clear() {
        this.cache.clear();
    }
    destroy() {
        clearInterval(this.timer);
        this.cache.clear();
    }
    evictExpired() {
        const now = Date.now();
        for (const [id, entry] of this.cache) {
            if (entry.expiresAt && now > entry.expiresAt) {
                this.cache.delete(id);
            }
        }
    }
}
exports.L1Cache = L1Cache;
