// @jackclaw/memory/dist/store.js
"use strict";
// L2 个人记忆 — SQLite 持久化存储
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.L2Store = void 0;
const better_sqlite3_1 = __importDefault(require("better-sqlite3"));
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
const DEFAULT_DB_PATH = path_1.default.join(os_1.default.homedir(), '.jackclaw', 'memory.db');
class L2Store {
    db;
    constructor(dbPath = DEFAULT_DB_PATH) {
        const dir = path_1.default.dirname(dbPath);
        import('fs').then(fs => fs.mkdirSync(dir, { recursive: true })).catch(() => { });
        // sync mkdir
        const fs = require('fs');
        fs.mkdirSync(dir, { recursive: true });
        this.db = new better_sqlite3_1.default(dbPath);
        this.db.pragma('journal_mode = WAL');
        this.db.pragma('synchronous = NORMAL');
        this.init();
    }
    init() {
        this.db.exec(`
      CREATE TABLE IF NOT EXISTS memories (
        id TEXT PRIMARY KEY,
        agentId TEXT NOT NULL,
        layer TEXT NOT NULL DEFAULT 'L2',
        category TEXT NOT NULL,
        scope TEXT NOT NULL DEFAULT 'private',
        content TEXT NOT NULL,
        tags TEXT NOT NULL DEFAULT '[]',
        importance REAL NOT NULL DEFAULT 0.5,
        createdAt INTEGER NOT NULL,
        updatedAt INTEGER NOT NULL,
        expiresAt INTEGER,
        source TEXT
      );
      CREATE INDEX IF NOT EXISTS idx_agent ON memories(agentId);
      CREATE INDEX IF NOT EXISTS idx_category ON memories(category);
      CREATE INDEX IF NOT EXISTS idx_scope ON memories(scope);
      CREATE INDEX IF NOT EXISTS idx_importance ON memories(importance DESC);
    `);
    }
    save(entry) {
        const stmt = this.db.prepare(`
      INSERT OR REPLACE INTO memories
        (id, agentId, layer, category, scope, content, tags, importance, createdAt, updatedAt, expiresAt, source)
      VALUES
        (@id, @agentId, @layer, @category, @scope, @content, @tags, @importance, @createdAt, @updatedAt, @expiresAt, @source)
    `);
        stmt.run({
            ...entry,
            tags: JSON.stringify(entry.tags),
            expiresAt: entry.expiresAt ?? null,
            source: entry.source ?? null,
        });
    }
    get(id) {
        const row = this.db.prepare('SELECT * FROM memories WHERE id = ?').get(id);
        return row ? this.rowToEntry(row) : undefined;
    }
    query(agentId, opts = {}) {
        const conditions = ['agentId = ?'];
        const params = [agentId];
        if (opts.layer) {
            conditions.push('layer = ?');
            params.push(opts.layer);
        }
        if (opts.category) {
            conditions.push('category = ?');
            params.push(opts.category);
        }
        if (opts.scope) {
            conditions.push('scope = ?');
            params.push(opts.scope);
        }
        if (opts.minImportance !== undefined) {
            conditions.push('importance >= ?');
            params.push(opts.minImportance);
        }
        const where = conditions.join(' AND ');
        const limit = opts.limit ?? 100;
        const rows = this.db.prepare(`SELECT * FROM memories WHERE ${where} ORDER BY importance DESC, updatedAt DESC LIMIT ?`).all([...params, limit]);
        let entries = rows.map(r => this.rowToEntry(r));
        if (opts.tags?.length) {
            entries = entries.filter(e => opts.tags.some(t => e.tags.includes(t)));
        }
        return entries;
    }
    delete(id) {
        this.db.prepare('DELETE FROM memories WHERE id = ?').run(id);
    }
    all(agentId) {
        const rows = this.db.prepare('SELECT * FROM memories WHERE agentId = ?').all(agentId);
        return rows.map(r => this.rowToEntry(r));
    }
    rowToEntry(row) {
        return {
            id: row['id'],
            agentId: row['agentId'],
            layer: row['layer'],
            category: row['category'],
            scope: row['scope'],
            content: row['content'],
            tags: JSON.parse(row['tags']),
            importance: row['importance'],
            createdAt: row['createdAt'],
            updatedAt: row['updatedAt'],
            expiresAt: row['expiresAt'],
            source: row['source'],
        };
    }
}
exports.L2Store = L2Store;
